### DISCUSSION TITLE
Progress towards version 1.1.0

### DISCUSSION THREAD

**Discussion by darianbjohnson** [Phambili-Tech/Newt_Display#20] [General]
Thought I would share what I've been working on for the last week (Aug 1 - Aug 8 2022).

![FZgGkJSWYAI6eCn](https://user-images.githubusercontent.com/21129686/183268185-331f834e-820b-4ccb-ac6a-1508f1faa218.jpg)

![FZgGkJTXoAIZ1YW](https://user-images.githubusercontent.com/21129686/183268191-9f210e9e-8682-40c0-8247-8bd184a2620b.jpg)

**Newt embedded software:**
- Adding option for uses to invert the colors of the screen - Light mode (default), Dark mode, and Auto mode (dark mode between 11PM - 5AM)
- Option to change between 12 and 24 hour time
- Release Notes/What's new included so people know what's new/changes in a version
- Option: Allowing users to switch between Imperial and Metric (will default to units based on location - user can override)
- Setting option for for a factory reset (removing all the settingss and wifi credentials)
- Fixing the WiFi provisioning (ESP Arduino 2.0.4 broke the Wifi Provisioning, which is why I had to push a hot fix last week)
- Working on weather API calls to get more weather details

**Newt Server Side Software**
- Evaluating different weather providers to find the API that provides the most information in the least number of calls. Each Newt makes ~50 weather calls per day... that scale pretty quickly across the number of Newts in the wild. Trying to get ahead of it before costs get out of control
- Evaluating a new MQTT bridge. I use one today that is good, but not as secure as I would like. I'd like to find something with increase security before I add user-identifying options to Newt (calendar, health, etc).

**Comment by tjfoth** [comment]
Dude, you've been busy!

I can't say a lot about cost of weather services, but what I've seen a lot of apps do to get around cost is to have each user get their own API  key.  Most weather services have free low volume API keys.

I'm currently very bullish on tomorrow.io, BTW.

By  sense is that  instead of getting data from a central server, you may need to distribute at least some of the data retrieval to each of the Newts.

Maybe you can support two or three weather service like OpenWeatherMap, tomorrow.io, and weather underground each from the  Newt.

I think you'll need to  allow for local/personal  MQTTs brokers, too.

**Comment by andypiper** [comment]
Wow, the new screen previews look lovely - nice work on those.

**Comment by angusmci** [comment]
Any chance that the end-of-Pomodoro sound can be made a little less terrifying? I realize that there are probably limits to what you can do with the Newt's speaker, but if there's a way to replace the loud and piercing shriek with something more like a gentle chirp, that would get my vote.

**Reply by darianbjohnson** [reply]
I logged an issue to track this. - [33](https://github.com/Phambili-Tech/Newt_Display/issues/33)

**Comment by badslug** [comment]
For weather data, you could try https://pirateweather.net for an open source option (and the maintainer of that could just be good collaborator and nicer place to spend your weather budget). 

Apple also has the new WeatherKit https://developer.apple.com/weatherkit/ with very high free limits if you have an apple developer account already.

**Reply by darianbjohnson** [reply]
Wow - this is great. I spent the last weekend  re-writing the backend to use Weatherapi.com (which costs $4 a month - totally reasonable)... but I've been unhappy with the weather hourly forecast. The pirateweather.net looks like a great option (and I'd happily donate $10 a month to the owner). The Apple Developer kit is a good option too. I've been thinking of signing up for a dev kit ($99/year)... that would give be access to weather AND other APIs for future use.

**Reply by badslug** [reply]
Yeah, it may not be a bad idea to have multiple potential weather choices and then move your usage or fallback based on things like the user's location (I think pirate weather's data is esp strong in certain regions and less so in others since it relies on mostly government public data) or other factors.

Is one of your goals for Newt to be sort of a more out-of-the-box information display device? My mind keeps jumping back to the chumby device if you are familiar with those. Some of these ideas feel a little over the top for more of a maker dev platform but would be a strong foundation for something like a chumby style device.

**Reply by darianbjohnson** [reply]
So - truth be told - I wasn't exactly sure what Newt was going to be. I really just wanted to get it out in the community to see where it went. I do think a "Pro" version (sold as a consumer product, with a proper case, supporting app, etc) would be kinda cool... but I think that's at least another hardware version away (the biggest drawback to newt today is that I can't keep a more persistent wifi connection to get realtime information... though there are a few new Wifi chips coming out that promise realtime connections without significant battery drainage).

**Comment by badslug** [comment]
Oh, yeah, and it's pretty niche but I have (and love) my Tempest weather station with has an API which would be extremely cool to see support for https://weatherflow.com/tempest-weather-system/

**Reply by darianbjohnson** [reply]
Interesting.. I've been thinking about a "store" that lets users add screens that might be niche (e.g. a fitbit integration, or google health integration... or an integration to the Tempest Weather System).... I'm going to open a new discussion thread to see if people have ideas about how to implement.

**Reply by badslug** [reply]
That sounds great. It would be awesome if there was some way to provide that level of customization without having to compile our own fork of the firmware. I think the ideal would be something that provides ways to customize data sources, as well as visual customization like the way Apple watch works with "faces" that have spaces for "complications" that you can plug in different views. So different health (fitbit, etc) or weather (open weather, tempest, etc) data sources could feed into several options for visualizing health or weather (e.g. plug in the large current temperature complication into this spot of this Newt face layout).

**Comment by tjfoth** [comment]
I like the idea of faces, but from a practical standpoint, it seems like
there has to be some limits in how data is fetched by the Newt.  I had
thought that perhaps it might be MQTT.  Maybe a simple URL/REST interface
expecting a JSON returned by the service would be adequate.

If you want something really flexible, it might make sense to do something
creative with Google Sheets.  That would allow people to write app scripts
to get data from a data source and put it into a Google Sheet for the Newt
to fetch it from.

I am wondering what other "free" services could be used as an
"intermediary" between a data service and the Newt that would be easy for
people to program?

---
Tom Foth
***@***.***
***@***.***

**Reply by badslug** [reply]
It does seem like an intermediate server that can do a lot of pre-processing would be super handy.

**Reply by angusmci** [reply]
The "intermediate server" for pre-processing idea makes a lot of sense. And while tying the Newt to Google doesn't entirely appeal, it's true that Google Sheets is fairly easy to work with, and likely to be around for a fairly long time. Github might be another candidate, on similar grounds.

If the Newt has or will have dependencies on particular remote servers to provide data, can we ensure that (a) the APIs are fully documented, and (b) the URLs are easily configurable. I've had bad experiences in the past with hardware that was hard-coded to connect to a specific server, and the company maintaining that server went out of business or abandoned the product.

"Intermediate" servers as suggested by @tjfoth and @badslug might be useful not just to do pre-processing but also to provide a consistent API. If you use data from `WeatherServiceFoo`, but you consume it via `MyNewtAPIServer` using the Newt Weather API, then when `WeatherServiceFoo` suddenly drops off the map, you don't need to hurriedly push out a firmware fix so that the Newt can talk to `WeatherServiceBar`. You just fix up `MyNewtAPIServer` to consume data from `WeatherServiceBar`. There are probably also advantages to offloading the heavy lifting involved in processing -- and caching -- the data on the server rather than the Newt.

Similarly, if you have configurable URLs and a standard API, anyone who likes `WeatherServiceBaz` better than `WeatherServiceFoo` can stand up their own API server, write their own transformer code to map from `WeatherServiceBaz` to Newt Weather API, and point their Newt at their own server instead.

**Reply by badslug** [reply]
@angusmci I completely agree about documenting the API extensively and making it easily reconfigurable. I've had the exact same experience with backed services getting abandoned and me getting stuck with essentially worthless hardware.

Maybe if we design the intermediate server using a common serverless backend (CloudFlare Workers etc) then it would be pretty trivial for people to fork the actual source code in use on the "official" API and deploy it to their own instance with their own tweaks to it. Serverless should be almost free for anyone to deploy/operate as an individual while still scaling for many users for the official deployment.

I do also really like offloading all the processing to the backend. Not only can we pre-process all the weather data so it's the same format no matter the source, but also perform all the conversions (e.g. to the preferred units: C/F, 12/24 hour time, etc) so Newt code can be as simple and reliable as possible.

**Comment by darianbjohnson** [comment]
Version 1.1.0 RC 1 to be released with in the next 24 hours! This won't be downloadable via Newt (yet) - right now I'm only releasing it as a manually downloadable file (along with the updated source code). Goal is to get the request documentation completed in the next two weeks, get user feedback, then make the code GA

**Comment by tjfoth** [comment]
I haven't had the time to build this version and try it yet (I just
installed a new sound system for church and a new video recording system),
so I'll be excited to try this new manually installable prebuilt version.

Regarding the intermediary server:

   - I fully get that Google is not everyone's first choice.  It isn't mine
   either.
   - That being said, there's a lot of code out there for supporting I/O to
   Google Sheets and so this seems like it makes sense.
   - Not only that, but writing in Google App Script (basically Javascript)
   doesn't completely suck. It's fine.
   - Everyone gets the spreadsheet metaphor
   - It is easy to manage sharing (you don't need credentials: you can
   share with a link).
   - I think there are some pretty easy things you could do to create
   server specified displays in the form of odd number cells in a row being
   labels and even numbered adjacent cells being values.

This would allow a lot of potential personalization of data and
presentation.


---
Tom Foth
***@***.***
***@***.***

**Comment by darianbjohnson** [comment]
I've started a new thread for discussion on the APIs and an "intermediate processing server" (which NEWT actually uses today). Documentation will evolve over time but a good start is here: 
https://github.com/Phambili-Tech/Newt_Display/wiki/MQTT-API-Summary

Includes link to source code, and descriptions of each API (and the source system for the API)

**Comment by koocotte** [comment]
It looks like 1.1.0 is out. In "settings", my newt say that it's the latest, but I do not have a button to upgrade. Is something wrong with my board? Will it come shortly? Or do I have to update manually with an USB-C wire?

**Reply by william-stearns** [reply]
I ended up upgrading it to 1.1.0 manually.  The instructions are at https://github.com/Phambili-Tech/Newt_Display/wiki/Upgrades .  My only note is that you need to copy the file very soon after Newt boots or the upgrade doesn't seem to happen.

**Reply by darianbjohnson** [reply]
Yeah, that's a known issue... I've been thinking about that a bit Simple answer/resolution might be to just document this as a known issue

**Reply by darianbjohnson** [reply]
@koocotte hmm - that's weird... what version does you board say you have?

**Reply by koocotte** [reply]
I have 1.0.2
![IMG_20221003_083856](https://user-images.githubusercontent.com/195378/193515354-53163931-855b-4491-9463-7a4e3f947623.jpg)

**Reply by darianbjohnson** [reply]
Well... that's a problem. I'm wondering if there is a bug in the code that doesn't recognize that v 1.1.0 is greater than 1.0.2. Can you do a favor for me, and restart your Newt (reset button or just turn in on/off). From there, can you let me know if you see a download option?

**Reply by william-stearns** [reply]
I had the same scenario as koocotte; current version was 1.0.2, "lastest" was 1.1.0, and no download option showed up in settings even after a reset.

**Reply by darianbjohnson** [reply]
Ugh - then that means there is a bug (sad trombone...). Ok, I think a pushing a new version 1.1.3, should (hopefully) solve the issue.... I'll add an issue

**Reply by koocotte** [reply]
Same as William, no download option after a reboot.

Do you plan to push 1.1.3 shortly and want us to test the regular upgrade? Or should we do a manual upgrade with cable?

**Reply by darianbjohnson** [reply]
Ok, can you hold off on manual updates for a day? I’ll bump the current version 1.1.0 to 1.1.3 tonight….hoping that fixes the issue

**Reply by koocotte** [reply]
I just rebooted the board. No change in the "setting" page; no upgrade button, current is 1.0.2 and latest 1.1.0

**Reply by darianbjohnson** [reply]
I wasn’t able to “bump up” the version number last night (was a bit under the weather). Goal is to do it tonight; I’ll update this thread when I’ve made the change.

**Reply by darianbjohnson** [reply]
Ok, I just updated the codebase to v 1.1.3 (and I made a hotfix to resolve the versioning issues). You should see an icon for download within the next hour.

**Reply by koocotte** [reply]
I got it and upgraded to 1.1.3. Thanks.

**Comment by william-stearns** [comment]
Nice work on 1.1.0!  I especially like the fact that weather can be on the home screen all the time instead of having to switch to it.

**Comment by william-stearns** [comment]
Perfect - after a reset I see a download icon in the upper left, the info
screen shows new version available 1.1.3, current 1.1.0, available 1.1.3,
and the left button shows upgrade.  Choosing that downloads 1.1.3  and
successfully reboots into it.  Thank you!
